enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,164 @@
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+### Requirements:
+
+1. CLI author requirements:
+    - provide CLI image containing binaries
+    - Custom Resource that describes the binaries (location within the image, metadata, description)
+2. An operator that will be responsible for:
+    - registering the new CLI CRD 
+    - reading metadata from CR from authors and extracting the binaries from the CLI images to local disk
+    - serving extracted binaries
+    - serving krew-compatible index
+3. oc changes:
+    - embed krew
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI in-cluster manager for various CLIs available for OpenShift and related services so that I can discover, install and list them
+The user will invoke the following commands:
+
+* `oc krew list` to view installed plugins
+* `oc krew search` will communicate with the krew index served by the cli-in-cluster-manager operator and will retrieve information about CLIs
```

Comment:
This has to be reviewed by the security team since you are adding a binary distribution mechanism.  Do you really need an operator / controller for this?  Why not just have the one CLI that will use krew calculate it from cluster CRs?

Reply:
For security review you can ask Jason Shepherd who has time to review your proposal.  You'll need to cover authenticity, transport verification, what on system permissions you'll need to perform this action.

This design needs to contrast to using RPMs directly and describe why you have bypassed that.  You'll need to describe the set of artifacts that you *must* support (i.e. the ones from products) and distinguish those from community / unsupported artifacts somehow.

Creating new binary distribution mechanisms has a pretty high bar at Red Hat.

Reply:
> This design needs to contrast to using RPMs directly and describe why you have bypassed that. You'll need to describe the set of artifacts that you _must_ support (i.e. the ones from products) and distinguish those from community / unsupported artifacts somehow.

The design is to provide access to compressed files, like the current consoleclidownloads, and use krew to install from those. It's taking what we have, with consoleclidownloads but then installing, and also with the ability to do that disconnected by using mirrored image registry.  If installing offline is not a requirement, I think the krew-index does not need to be served in-cluster.  Instead, we can maintain a GH repo like `openshift/oc-plugin-index` or something, and configure oc with that index.  I was under the impression that disconnected binary installs was a main requirement.  The proposal can be simplified, for sure, if this is not the case.